api/test-client.js

```diff
@@ -27,20 +53,24 @@ tap.test('connect', t => {
     ws.close('bad wolf');
   });
 
+  function validateLoginFailure(t, error, conn) {
+    helpers.requestEqual(t, ws.lastRequest, {
+      type: 'Admin',
+      request: 'Login',
+      params: {'auth-tag': 'who', credentials: 'secret', macaroons: []},
+      version: 3
+    });
+    t.equal(error, 'bad wolf');
+    t.equal(conn, null);
+    t.end();
```

Comment:
Thanks for abstracting this out, but I'd leave the responsibility of calling t.end() to the actual test.